workflow.py

```diff
@@ -214,54 +215,114 @@ async def eliminate_subtraction(
+    # as to not disrupt uses elsewhere
+    shutil.copyfile(isolate_sam_path, working_isolate_path)
 
-    await aiofiles.os.remove(to_subtraction_sam_path)
+    # copy the original fastq file into a working fastq file
+    # as to not disrupt uses elsewhere
+    shutil.copyfile(isolate_fastq_path, working_fastq_path)
 
     subtracted_count = 0
 
-    async with aiofiles.open("subtracted_read_ids.txt", "r") as f:
-        async for line in f:
-            if line:
-                subtracted_count += 1
-
-    logger.info(
-        f"Subtracted {subtracted_count} reads that mapped better to a subtraction."
-    )
+    # iterate over all subtractions
+    for subtraction in subtractions:
+        # Map reads to the subtraction.
+        await run_subprocess(
+            [
+                "bowtie2",
+                "--local",
+                "--no-unal",
+                "--no-hd",
+                "--no-sq",
+                "-N",
+                "0",
+                "-p",
+                str(proc),
+                "-x",
+                shlex.quote(str(subtraction.bowtie2_index_path)),
+                "-U",
+                str(working_fastq_path),
+                "-S",
+                str(to_subtraction_sam_path),
+            ]
+        )
+
+        await run_subprocess(
+            [
+                "./eliminate_subtraction",
+                str(working_isolate_path),
+                str(to_subtraction_sam_path),
+                str(subtracted_sam_path),
+            ]
+        )
+
+        await aiofiles.os.remove(to_subtraction_sam_path)
+
+        async with aiofiles.open("subtracted_read_ids.txt", "r") as f:
+            async for line in f:
+                if line:
+                    subtracted_count += 1
+
+        # rewrite the fastq file to exclude previous reads
+
+        lines = None
+        with open(working_fastq_path, 'r') as f:
+            lines = f.readlines()
+
+        # group every 4 lines of isolate_fastq into a collection
+        grouped_fastq_reads = []
+        i = 0
+        while (i + 3) < len(lines):
+            try:
+                grouped_fastq_reads.append(
+                    [
+                        lines[i],
+                        lines[i + 1],
+                        lines[i + 2],
+                        lines[i + 3]
+                    ]
+                )
+                i += 4
+            except IndexError:
+                break
+
+        # load in the previously subtracted data
+        subtracted_reads = []
+        with open("subtracted_read_ids.txt", "r") as f:
+            subtracted_reads = [line.strip() for line in f]
+
+        # unwrap collections and rewrite into new isolate_fastq file;
+        # only write if not previously subtracted
+        with open(working_fastq_path, 'w') as f:
+            for read in grouped_fastq_reads:
+                if read[0].strip("@\n") not in subtracted_reads:
```

Comment:
Always use `set` for checks like this. Using `in` with a list means traversing the entire data structure, which in this case is probably quite long.

https://www.geeksforgeeks.org/sets-in-python/